full-node/sov-stf-runner/src/prover_service/parallel/prover.rs

```diff
@@ -0,0 +1,178 @@
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use sov_rollup_interface::da::{BlockHeaderTrait, DaSpec};
+use sov_rollup_interface::services::da::DaService;
+use sov_rollup_interface::stf::StateTransitionFunction;
+use sov_rollup_interface::zk::{Proof, ZkvmHost};
+
+use super::{Hash, ProverServiceError};
+use crate::{ProofGenConfig, ProofSubmissionStatus, StateTransitionData};
+
+enum ProverStatus<StateRoot, Witness, Da: DaSpec> {
+    WitnessSubmitted(StateTransitionData<StateRoot, Witness, Da>),
+    Proving,
+    Proved(Proof),
+    Err(anyhow::Error),
+}
+
+#[derive(Default)]
+struct ProverState<StateRoot, Witness, Da: DaSpec> {
+    prover_status: HashMap<Hash, ProverStatus<StateRoot, Witness, Da>>,
+}
+
+impl<StateRoot, Witness, Da: DaSpec> ProverState<StateRoot, Witness, Da> {
+    fn remove(&mut self, hash: &Hash) -> Option<ProverStatus<StateRoot, Witness, Da>> {
```

Comment:
The lock is taken over `ProverState` in the `Prover`. It is done this way because in the following PR, I will add a flag related to the `busy` state management and this will need to be protected by the same lock.